File: api/gemini.js
// api/gemini.js
import axios from 'axios';

export default async function handler(req, res) {
    // 1. Cấu hình CORS
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'POST, OPTIONS');
    res.setHeader('Access-Control-Allow-Headers', 'Content-Type');

    // 2. Xử lý preflight
    if (req.method === 'OPTIONS') {
        return res.status(200).end();
    }

    // 3. Chỉ chấp nhận POST
    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method not allowed. Use POST.' });
    }

    // 4. Kiểm tra body
    if (!req.body || typeof req.body !== 'object') {
        return res.status(400).json({ 
            error: 'Invalid request body. Ensure Content-Type is application/json and body is valid JSON.' 
        });
    }

    try {
        // Lấy dữ liệu từ request của userscript
        const { model = 'gemini-2.5-flash', messages, temperature = 0.1, max_tokens = 50 } = req.body;

        if (!messages || !Array.isArray(messages)) {
            return res.status(400).json({ error: 'Missing or invalid "messages" field.' });
        }

        // Xây dựng prompt theo định dạng của Gemini API
        // Gemini dùng mảng 'parts' thay vì 'messages' của OpenAI
        const contents = messages.map(msg => ({
            role: msg.role === 'assistant' ? 'model' : 'user', // Gemini dùng 'model' thay cho 'assistant'
            parts: [{ text: msg.content }]
        }));

        // 5. Gọi Gemini API
        const response = await axios.post(
            `https://generativelanguage.googleapis.com/v1beta/models/${model}:generateContent`,
            {
                contents: contents,
                generationConfig: {
                    temperature: temperature,
                    maxOutputTokens: max_tokens
                }
            },
            {
                headers: {
                    'Content-Type': 'application/json',
                    'x-goog-api-key': process.env.GEMINI_API_KEY // 🔐 Sử dụng biến môi trường
                }
            }
        );

        // 6. Chuẩn hóa response về định dạng giống OpenAI để userscript không cần sửa nhiều
        const geminiResponse = response.data;
        const candidate = geminiResponse.candidates?.[0];
        const answerText = candidate?.content?.parts?.[0]?.text || "No response";

        // Tạo response có cấu trúc giống OpenAI
        const openAIFormattedResponse = {
            choices: [{
                message: {
                    role: 'assistant',
                    content: answerText.trim()
                }
            }]
        };

        res.status(200).json(openAIFormattedResponse);
    } catch (error) {
        console.error('Proxy error:', error.response?.data || error.message);
        res.status(error.response?.status || 500).json({
            error: 'Failed to fetch from Gemini API',
            details: error.response?.data || error.message
        });
    }
}

File: api/gemini.py
# api/gemini.py
from flask import Flask, request, jsonify
from flask_cors import CORS
import requests
import os
import time
import logging

app = Flask(__name__)
CORS(app)

GEMINI_API_KEY = os.environ.get("GEMINI_API_KEY")
if not GEMINI_API_KEY:
    raise ValueError("Thiếu biến môi trường GEMINI_API_KEY")

GEMINI_MODEL = "gemini-2.5-flash" # Bạn có thể đổi model tại đây
MAX_RETRIES = 3

# Thiết lập logging cơ bản để dễ dàng theo dõi lỗi
logging.basicConfig(level=logging.INFO)

@app.route("/api/gemini", methods=["POST", "OPTIONS"])
def proxy_gemini():
    if request.method == "OPTIONS":
        return "", 200

    try:
        data = request.get_json(force=True)
        messages = data.get("messages", [])
        temperature = data.get("temperature", 0.1)
        max_tokens = data.get("max_tokens", 50)

        contents = []
        for msg in messages:
            role = "user"
            if msg.get("role") == "assistant":
                role = "model"
            contents.append({
                "role": role,
                "parts": [{"text": msg.get("content", "")}]
            })

        url = f"https://generativelanguage.googleapis.com/v1beta/models/{GEMINI_MODEL}:generateContent"
        headers = {"Content-Type": "application/json", "x-goog-api-key": GEMINI_API_KEY}
        payload = {
            "contents": contents,
            "generationConfig": {
                "temperature": temperature,
                "maxOutputTokens": max_tokens
            }
        }

        answer_text = ""
        for attempt in range(MAX_RETRIES):
            try:
                resp = requests.post(url, headers=headers, json=payload, timeout=8.0)
                resp.raise_for_status()
                gemini_data = resp.json()

                # Kiểm tra và trích xuất câu trả lời
                if "candidates" in gemini_data and len(gemini_data["candidates"]) > 0:
                    candidate = gemini_data["candidates"][0]
                    
                    # Kiểm tra xem phản hồi có bị chặn bởi bộ lọc an toàn không
                    if candidate.get("finishReason") == "SAFETY":
                        logging.warning(f"Lần thử {attempt+1}: Phản hồi bị chặn bởi bộ lọc an toàn.")
                        time.sleep(2 ** attempt) # Exponential backoff
                        continue

                    if "content" in candidate and "parts" in candidate["content"]:
                        answer_text = candidate["content"]["parts"][0].get("text", "").strip()
                        if answer_text:
                            break # Thành công, thoát vòng lặp
                        else:
                            logging.warning(f"Lần thử {attempt+1}: Nhận được nội dung rỗng. Đang thử lại...")
                            time.sleep(2 ** attempt)
                else:
                    logging.warning(f"Lần thử {attempt+1}: Không tìm thấy 'candidates' trong phản hồi.")
                    time.sleep(2 ** attempt)

            except requests.exceptions.RequestException as e:
                logging.error(f"Lần thử {attempt+1}: Lỗi mạng hoặc API: {e}")
                time.sleep(2 ** attempt)

        # Trả về kết quả cuối cùng (có thể là rỗng nếu đã thử hết số lần)
        return jsonify({
            "choices": [{
                "message": {
                    "role": "assistant",
                    "content": answer_text
                }
            }]
        })

    except Exception as e:
        logging.error(f"Lỗi proxy không mong muốn: {e}")
        return jsonify({"error": str(e)}), 500
